File: bench/programs/unions/unions.go
package main

import (
	"fmt"
	"time"
)

// Sum-type via interface + type switch (Go's idiom for discriminated unions).
type Expr interface{ isExpr() }

type Num struct{ v int64 }
type Add struct{ l, r Expr }
type Mul struct{ l, r Expr }

func (Num) isExpr() {}
func (Add) isExpr() {}
func (Mul) isExpr() {}

func evaluate(e Expr) int64 {
	switch x := e.(type) {
	case Num:
		return x.v
	case Add:
		return evaluate(x.l) + evaluate(x.r)
	case Mul:
		return evaluate(x.l) * evaluate(x.r)
	}
	return 0
}

func run() int64 {
	var count, p int64 = 2000000, 1000000007
	var acc int64 = 0
	for i := int64(0); i < count; i++ {
		var expr Expr = Add{
			l: Num{v: i % 1000},
			r: Mul{l: Num{v: i % 7}, r: Num{v: 3}},
		}
		acc = (acc + evaluate(expr)) % p
	}
	return acc
}

func main() {
	run() // warmup
	t0 := time.Now()
	ans := run()
	dt := time.Since(t0).Nanoseconds()
	fmt.Println(ans)
	fmt.Println(dt)
}

File: bench/programs/unions/unions.js
// Tagged-object discriminated union + switch.
const NUM = 0, ADD = 1, MUL = 2;

function evaluate(e) {
  switch (e.t) {
    case NUM: return e.v;
    case ADD: return evaluate(e.l) + evaluate(e.r);
    case MUL: return evaluate(e.l) * evaluate(e.r);
  }
}

function run() {
  const count = 2000000, p = 1000000007;
  let acc = 0;
  for (let i = 0; i < count; i++) {
    const expr = {
      t: ADD,
      l: { t: NUM, v: i % 1000 },
      r: { t: MUL, l: { t: NUM, v: i % 7 }, r: { t: NUM, v: 3 } },
    };
    acc = (acc + evaluate(expr)) % p;
  }
  return acc;
}

run(); // warmup the JIT
const t0 = process.hrtime.bigint();
const ans = run();
const dt = process.hrtime.bigint() - t0;
console.log(ans.toString());
console.log(dt.toString());
